### PR TITLE
Import: Change merge logic

### LINKED ISSUE #6162: Bug: Import should replace the destination prefix with the source's state
Currently, Importing into a lakeFS branch will result in a union of both source and destination, with the source taking precedence. This means that If I have:

```text
source:
    - 1.txt (identity: abc)
    - 2.txt (identity: def)
    - 3.txt (identity: ghi)

destination:
    - 1.txt (identity: jkl)
    - 4.txt (identity: mno)
```

**Current behavior** 

Importing source into destination will result in:

```text
destination:
    - 1.txt (identity: abc)  # notice the source's identity, because source-wins
    - 2.txt (identity: def)
    - 3.txt (identity: ghi)
    - 4.txt (identity: mno)
```

**Expected behavior**

```text
destination:
    - 1.txt (identity: abc)
    - 2.txt (identity: def)
    - 3.txt (identity: ghi)

```

The reason for this to be the expected behavior is that merging these states across 2 diverging history lines (changes applied in source vs changes applied in destination) will typically result in a logical AND. 

This is because most distributed compute frameworks (Spark but also others) will include some random string in the object name, so objects are rarely overwritten. Even if we apply "source-wins", any change made on the destination will not be removed in favor of the changes made on source.

cc @N-o-Z

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
In order to support some cases of the import behavior, additional checks were needed to be added to the `merger` logic.
This PR includes those changes and tests for the import behavior.

It's recommended to review #6250 (the `PrefixIterator` is sufficient)

Closes #6162

### CODE REVIEW

File: pkg/graveler/committed/merge.go [hunk 1]
```diff
@@ -229,7 +229,7 @@ func (m *merger) handleBothRanges(sourceRange *Range, destRange *Range) error {
 		m.haveSource = m.source.NextRange()
 		m.haveDest = m.dest.NextRange()
 
-	case bytes.Equal(sourceRange.MinKey, destRange.MinKey) && bytes.Equal(sourceRange.MaxKey, destRange.MaxKey): // same bounds
+	case sourceRange.EqualBounds(destRange):
```
Comment (johnnyaug) [inline pkg/graveler/committed/merge.go:232]: Nice!!

**Comment by johnnyaug** [comment]
Like I said in #6250, while it's great to reuse the merge logic for the import, I don't think the import is part of the merge.
This PR makes this a bigger issue. 

 I feel responsible for taking us down this path but I can only suggest two things:
1. Try to solve this by adding the following logic to the previous PR: the destination iterator should split any range that (may) include a prefix. For example, if the prefix is b and the original range is [a,c], that dest iterator will return [a,b] and [b,c]. This may solve the relevant issue and still will allow you to leave the merge code as is.
2. Go back to a dedicated import solution. :(

File: pkg/graveler/committed/merge.go [hunk 2]
```diff
@@ -370,7 +370,8 @@ func (m *merger) handleDestRangeSourceKey(destRange *Range, sourceValue *gravele
 		return m.sourceBeforeDest(sourceValue)
 	}
 
-	if bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0 { // dest range before source
+	if bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0 &&
+		(m.strategy != graveler.MergeStrategyImport || m.validImportRange()) { // dest range before source
```
Comment (nopcoder) [inline pkg/graveler/committed/merge.go:374]: graveler merge code should not address special import rage by casting iterator implementation. the solution should be a general solution that works with any merge or by provide merge options that will customize the merge we try to apply, but by graveler merge knowing this is merge.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/graveler/committed/range.go [hunk 3]
```diff
@@ -23,6 +27,14 @@ func (r Range) Copy() *Range {
 	}
 }
 
+func (r Range) EqualBounds(o *Range) bool {
```
[the remark this answers is not in the document]
Comment (Jonathan-Rosenberg) [reply pkg/graveler/committed/range.go:30]: Since it's not preforming a full Range equality check, maybe `EqualRangeBounds`?

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		return false
+	}
+	vr, r := ipi.updateValue()
+	ipi.updatePrefix()
+
+	if vr == nil && r != nil { // head of range
+		for ipi.IsCurrentRangeBoundedByPrefix() {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:62]: Shouldn't happen.

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+}
+func (ipi *SkipPrefixIterator) SeekGE(id graveler.Key) {
+	ipi.rangeIterator.SeekGE(id)
+}
+
+func (ipi *SkipPrefixIterator) Next() bool {
+	hasNext := ipi.rangeIterator.Next()
+	if !hasNext {
+		return false
+	}
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:54]:
```suggestion
	if !ipi.rangeIterator.Next() {
		return false
	}
```
Also in other places.

File: pkg/graveler/committed/manager.go [hunk 6]
```diff
@@ -190,30 +201,47 @@ func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespa
...
-		return "", fmt.Errorf("get base iterator: %w", err)
-	}
-	defer baseIt.Close()
+	return c.merge(ctx, ns, destination, source, base, iterators{}, strategy)
+}
 
-	destIt, err := c.metaRangeManager.NewMetaRangeIterator(ctx, ns, destination)
-	if err != nil {
-		return "", fmt.Errorf("get destination iterator: %w", err)
+type iterators struct {
```
[the remark this answers is not in the document]
Comment (Jonathan-Rosenberg) [reply pkg/graveler/committed/manager.go]: done

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+func (ipi *SkipPrefixIterator) updatePrefix() {
+	if ipi.currentPrefixIndex >= len(ipi.prefixes) {
+		return
+	}
+	currMinKey := string(ipi.currentRangeValue.r.MinKey)
+	if ipi.currentRangeValue.vr != nil {
+		currMinKey = string(ipi.currentRangeValue.vr.Key)
+	}
+	// If the current prefix is smaller or isn't the prefix of the currentMinKey, get the next prefix.
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:117]:
```suggestion
	// If the current prefix is smaller and isn't the prefix of the currentMinKey, get the next prefix.
```
right?

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	// By the end of this loop, the examined prefix will either be the prefix of the currentMinKey, or
+	// lexicographically bigger than it.
+	for string(ipi.prefixes[ipi.currentPrefixIndex]) < currMinKey &&
+		!strings.HasPrefix(currMinKey, string(ipi.prefixes[ipi.currentPrefixIndex])) {
+		p := ipi.currentPrefixIndex + 1
+		switch {
+		case p >= len(ipi.prefixes): // No more comparable prefixes
+			return
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:125]:
```suggestion
			ipi.currentPrefixIndex = p
			return
```
So next time function will return fast?

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	if !hasNext {
+		return false
+	}
+	ipi.updateValue()
+	ipi.updatePrefix()
+
+	for ipi.IsCurrentRangeBoundedByPrefix() {
+		if ipi.rangeIterator.Err() != nil {
+			return false
+		}
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:98]: Shouldn't happen

File: pkg/graveler/committed/skip_prefix_iterator.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+		hasNext = ipi.rangeIterator.NextRange()
+		if !hasNext {
+			return false
+		}
+		ipi.updateValue()
+		ipi.currentPrefixIndex++
```
Comment (johnnyaug) [inline pkg/graveler/committed/skip_prefix_iterator.go:104]: Why? What if the next range is also bounded by the prefix?

Comment (Jonathan-Rosenberg) [reply pkg/graveler/committed/skip_prefix_iterator.go:104]: True
This should be like in `Next()` and call `updatePrefix`.

File: pkg/graveler/committed/manager.go [hunk 11]
```diff
@@ -180,6 +180,26 @@ func (c *committedManager) Diff(ctx context.Context, ns graveler.StorageNamespac
...
+	mctx := mergeContext{
+		destIt:        destIt,
+		srcIt:         nil,
+		baseIt:        nil,
+		strategy:      graveler.MergeStrategyNone,
+		ns:            ns,
+		destinationID: destination,
+		sourceID:      source,
+		baseID:        "",
+	}
```
Comment (nopcoder) [inline pkg/graveler/committed/manager.go:199]:
```suggestion
	mctx := mergeContext{
		destIt:        destIt,
		strategy:      graveler.MergeStrategyNone,
		ns:            ns,
		destinationID: destination,
		sourceID:      source,
	}
```

File: pkg/graveler/committed/manager.go [hunk 12]
```diff
@@ -189,43 +209,73 @@ func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespa
...
+	mctx := mergeContext{
+		destIt:        nil,
+		srcIt:         nil,
+		baseIt:        nil,
+		strategy:      strategy,
+		ns:            ns,
+		destinationID: destination,
+		sourceID:      source,
+		baseID:        base,
+	}
```
Comment (nopcoder) [inline pkg/graveler/committed/manager.go:221]:
```suggestion
	mctx := mergeContext{
		strategy:      strategy,
		ns:            ns,
		destinationID: destination,
		sourceID:      source,
		baseID:        base,
	}
```

File: pkg/graveler/committed/manager.go [hunk 13]
```diff
@@ -189,43 +209,73 @@ func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespa
...
+	return c.merge(ctx, mctx)
+}
 
-	baseIt, err := c.metaRangeManager.NewMetaRangeIterator(ctx, ns, base)
-	if err != nil {
-		return "", fmt.Errorf("get base iterator: %w", err)
+type mergeContext struct {
+	destIt, srcIt, baseIt           Iterator
```
Comment (nopcoder) [inline pkg/graveler/committed/manager.go:226]:
```suggestion
	baseIt                          Iterator		
	srcIt                           Iterator
	destIt                          Iterator
```

File: pkg/graveler/committed/manager.go [hunk 14]
```diff
@@ -189,43 +209,73 @@ func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespa
...
+	strategy                        graveler.MergeStrategy
+	ns                              graveler.StorageNamespace
+	destinationID, sourceID, baseID graveler.MetaRangeID
```
Comment (nopcoder) [inline pkg/graveler/committed/manager.go:229]:
```suggestion
	baseID graveler.MetaRangeID
	sourceID graveler.MetaRangeID
	destinationID graveler.MetaRangeID		
```

**Review by nopcoder — APPROVED** [review]
LGTM with minor code style suggestion

**Review by johnnyaug — APPROVED** [review]
Looks great!

**Pull request merged**